src/reflect/scala/reflect/internal/Types.scala

```diff
@@ -2789,18 +2814,28 @@ trait Types
        *  See SI-5359.
        */
       val bounds  = tparam.info.bounds
+
       /* We can seed the type constraint with the type parameter
        * bounds as long as the types are concrete.  This should lower
        * the complexity of the search even if it doesn't improve
        * any results.
        */
-      if (propagateParameterBoundsToTypeVars) {
-        val exclude = bounds.isEmptyBounds || (bounds exists typeIsNonClassType)
+      val constr =
+        if ((propagateParameterBoundsToTypeVars || settings.YliteralTypes) && !isPastTyper) {
```

Comment:
Can we move the other side of the disjunction to propagateParameterBoundsToTypeVars's rhs? Should also dig in why we're getting the SOE -- is it because of existential type subtyping?